File: rust/tests/session_name_status_test.rs
//! Tests for --status lookup by session name and detached status enrichment
//! Issue #101: --session name not usable with --status, and --detached reports immediate completion
//! Issue #105: currentTime added to --status output for executing commands

use start_command::{
    attach_current_time, enrich_detached_status, is_detached_session_alive, query_status,
    ExecutionRecord, ExecutionRecordOptions, ExecutionStatus, ExecutionStore,
    ExecutionStoreOptions,
};
use std::collections::HashMap;
use tempfile::TempDir;

/// Helper to create a test store in a temporary directory
fn create_test_store() -> (TempDir, ExecutionStore) {
    let temp_dir = TempDir::new().unwrap();
    let store = ExecutionStore::with_options(ExecutionStoreOptions {
        app_folder: Some(temp_dir.path().to_path_buf()),
        use_links: Some(false),
        verbose: false,
    });
    (temp_dir, store)
}

/// Helper to create isolation options with session name
fn make_isolation_options(
    session_name: &str,
    isolated: &str,
    isolation_mode: &str,
) -> HashMap<String, serde_json::Value> {
    let mut opts = HashMap::new();
    opts.insert(
        "sessionName".to_string(),
        serde_json::Value::String(session_name.to_string()),
    );
    opts.insert(
        "isolated".to_string(),
        serde_json::Value::String(isolated.to_string()),
    );
    opts.insert(
        "isolationMode".to_string(),
        serde_json::Value::String(isolation_mode.to_string()),
    );
    opts
}

// ===== ExecutionStore::get() session name lookup tests =====

#[test]
fn test_get_by_uuid() {
    let (_temp_dir, store) = create_test_store();

    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo hello".to_string(),
        uuid: Some("test-uuid-session-101".to_string()),
        pid: Some(12345),
        options: Some(make_isolation_options("my-session", "screen", "attached")),
        ..Default::default()
    });
    record.complete(0);
    store.save(&record).unwrap();

    let found = store.get("test-uuid-session-101");
    assert!(found.is_some());
    assert_eq!(found.unwrap().uuid, "test-uuid-session-101");
}

#[test]
fn test_get_by_session_name() {
    let (_temp_dir, store) = create_test_store();

    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 60".to_string(),
        uuid: Some("uuid-for-session-lookup-test".to_string()),
        pid: Some(12345),
        options: Some(make_isolation_options(
            "my-custom-session",
            "screen",
            "detached",
        )),
        ..Default::default()
    });
    store.save(&record).unwrap();

    let found = store.get("my-custom-session");
    assert!(found.is_some());
    let found = found.unwrap();
    assert_eq!(found.uuid, "uuid-for-session-lookup-test");
    assert_eq!(
        found.options.get("sessionName").unwrap().as_str().unwrap(),
        "my-custom-session"
    );
}

#[test]
fn test_get_prefers_uuid_over_session_name() {
    let (_temp_dir, store) = create_test_store();

    // Record 1 with a specific UUID
    let mut record1 = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo first".to_string(),
        uuid: Some("target-uuid-101".to_string()),
        pid: Some(111),
        options: Some(make_isolation_options("some-session", "screen", "attached")),
        ..Default::default()
    });
    record1.complete(0);
    store.save(&record1).unwrap();

    // Record 2 whose session name matches record1's UUID
    let record2 = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo second".to_string(),
        uuid: Some("other-uuid-101".to_string()),
        pid: Some(222),
        options: Some(make_isolation_options(
            "target-uuid-101",
            "screen",
            "detached",
        )),
        ..Default::default()
    });
    store.save(&record2).unwrap();

    // Looking up by record1's UUID should return record1, not record2
    let found = store.get("target-uuid-101").unwrap();
    assert_eq!(found.command, "echo first");
}

#[test]
fn test_get_nonexistent_session_name() {
    let (_temp_dir, store) = create_test_store();

    let found = store.get("nonexistent-session");
    assert!(found.is_none());
}

#[test]
fn test_get_record_without_session_name() {
    let (_temp_dir, store) = create_test_store();

    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo hello".to_string(),
        uuid: Some("no-session-name-uuid".to_string()),
        pid: Some(12345),
        ..Default::default()
    });
    store.save(&record).unwrap();

    let found = store.get("some-session-name");
    assert!(found.is_none());
}

// ===== query_status() with session name tests =====

#[test]
fn test_query_status_by_session_name() {
    let (_temp_dir, store) = create_test_store();

    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 60".to_string(),
        uuid: Some("query-session-uuid-101".to_string()),
        pid: Some(12345),
        options: Some(make_isolation_options(
            "my-query-session",
            "screen",
            "attached",
        )),
        ..Default::default()
    });
    record.complete(0);
    store.save(&record).unwrap();

    let result = query_status(Some(&store), "my-query-session", Some("json"));
    assert!(result.success);
    let output = result.output.unwrap();
    assert!(output.contains("query-session-uuid-101"));
    assert!(output.contains("sleep 60"));
}

#[test]
fn test_query_status_nonexistent_session_name() {
    let (_temp_dir, store) = create_test_store();

    let result = query_status(Some(&store), "nonexistent-session", Some("json"));
    assert!(!result.success);
    assert!(result
        .error
        .unwrap()
        .contains("No execution found with UUID or session name"));
}

// ===== Detached status enrichment tests =====

#[test]
fn test_is_detached_session_alive_non_detached() {
    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo hello".to_string(),
        options: Some(make_isolation_options("test", "screen", "attached")),
        ..Default::default()
    });
    assert!(is_detached_session_alive(&record).is_none());
}

#[test]
fn test_is_detached_session_alive_no_session_name() {
    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo hello".to_string(),
        ..Default::default()
    });
    assert!(is_detached_session_alive(&record).is_none());
}

#[test]
fn test_is_detached_session_alive_nonexistent_screen() {
    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 60".to_string(),
        options: Some(make_isolation_options(
            "nonexistent-screen-session-test-101",
            "screen",
            "detached",
        )),
        ..Default::default()
    });
    let alive = is_detached_session_alive(&record);
    // May be Some(false) or None depending on whether screen is installed
    if let Some(v) = alive {
        assert!(!v);
    }
}

#[test]
fn test_enrich_detached_status_non_detached() {
    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo hello".to_string(),
        options: Some(make_isolation_options("test", "screen", "attached")),
        ..Default::default()
    });
    record.complete(0);

    let enriched = enrich_detached_status(&record);
    assert_eq!(enriched.status, ExecutionStatus::Executed);
    assert_eq!(enriched.exit_code, Some(0));
}

#[test]
fn test_enrich_detached_status_marks_dead_session_as_executed() {
    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 60".to_string(),
        options: Some(make_isolation_options(
            "nonexistent-session-enrich-101",
            "screen",
            "detached",
        )),
        ..Default::default()
    });
    // Record says executing, but session doesn't exist

    let enriched = enrich_detached_status(&record);
    // If screen is available, should mark as executed with exit code -1
    if enriched.status == ExecutionStatus::Executed {
        assert_eq!(enriched.exit_code, Some(-1));
        assert!(enriched.end_time.is_some());
    }
}

#[test]
fn test_get_most_recent_session_name_match() {
    let (_temp_dir, store) = create_test_store();

    // Create two records with the same session name (e.g., reuse of session name)
    let mut record1 = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo first".to_string(),
        uuid: Some("older-uuid-101".to_string()),
        pid: Some(111),
        options: Some(make_isolation_options(
            "reused-session",
            "screen",
            "attached",
        )),
        ..Default::default()
    });
    record1.complete(0);
    store.save(&record1).unwrap();

    let record2 = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo second".to_string(),
        uuid: Some("newer-uuid-101".to_string()),
        pid: Some(222),
        options: Some(make_isolation_options(
            "reused-session",
            "screen",
            "detached",
        )),
        ..Default::default()
    });
    store.save(&record2).unwrap();

    // Should find the first matching record (order depends on storage)
    let found = store.get("reused-session");
    assert!(found.is_some());
    // Both records have this session name; get() returns the first match
    let found = found.unwrap();
    assert!(found.uuid == "older-uuid-101" || found.uuid == "newer-uuid-101");
}

// ===== Issue #105: attach_current_time for executing status =====

#[test]
fn test_attach_current_time_returns_some_for_executing_record() {
    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 60".to_string(),
        uuid: Some("issue-105-executing".to_string()),
        pid: Some(12345),
        status: Some(ExecutionStatus::Executing),
        log_path: Some("/tmp/test.log".to_string()),
        ..Default::default()
    });

    let before = chrono::Utc::now();
    let current_time = attach_current_time(&record);
    let after = chrono::Utc::now();

    assert!(current_time.is_some());
    let ct = current_time.unwrap();
    let parsed = chrono::DateTime::parse_from_rfc3339(&ct)
        .expect("currentTime must be a valid RFC3339 timestamp");
    assert!(parsed >= before - chrono::Duration::milliseconds(1));
    assert!(parsed <= after + chrono::Duration::milliseconds(1));
}

#[test]
fn test_attach_current_time_returns_none_for_executed_record() {
    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo hello".to_string(),
        uuid: Some("issue-105-executed".to_string()),
        pid: Some(12345),
        log_path: Some("/tmp/test.log".to_string()),
        ..Default::default()
    });
    record.complete(0);

    assert_eq!(record.status, ExecutionStatus::Executed);
    assert!(attach_current_time(&record).is_none());
}

#[test]
fn test_attach_current_time_does_not_mutate_record() {
    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 60".to_string(),
        uuid: Some("issue-105-no-mutation".to_string()),
        pid: Some(12345),
        status: Some(ExecutionStatus::Executing),
        log_path: Some("/tmp/test.log".to_string()),
        ..Default::default()
    });
    let snapshot = record.clone();
    let _ = attach_current_time(&record);
    assert_eq!(record.uuid, snapshot.uuid);
    assert_eq!(record.status, snapshot.status);
    assert_eq!(record.start_time, snapshot.start_time);
    assert_eq!(record.end_time, snapshot.end_time);
    assert_eq!(record.exit_code, snapshot.exit_code);
}

// ===== Issue #105: query_status surfaces currentTime via all formats =====

#[test]
fn test_query_status_json_includes_current_time_for_executing() {
    let (_temp_dir, store) = create_test_store();
    let before = chrono::Utc::now();

    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 100".to_string(),
        uuid: Some("issue-105-json-executing".to_string()),
        pid: Some(99999),
        status: Some(ExecutionStatus::Executing),
        log_path: Some("/tmp/executing.log".to_string()),
        ..Default::default()
    });
    store.save(&record).unwrap();

    let result = query_status(Some(&store), "issue-105-json-executing", Some("json"));
    assert!(result.success);
    let output = result.output.unwrap();
    let parsed: serde_json::Value = serde_json::from_str(&output).unwrap();

    let ct = parsed["currentTime"]
        .as_str()
        .expect("currentTime should be present and a string");
    let parsed_ct = chrono::DateTime::parse_from_rfc3339(ct)
        .expect("currentTime must be a valid RFC3339 timestamp");
    let after = chrono::Utc::now();
    assert!(parsed_ct >= before - chrono::Duration::seconds(1));
    assert!(parsed_ct <= after + chrono::Duration::seconds(1));
}

#[test]
fn test_query_status_json_omits_current_time_for_executed() {
    let (_temp_dir, store) = create_test_store();

    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo done".to_string(),
        uuid: Some("issue-105-json-executed".to_string()),
        pid: Some(11111),
        log_path: Some("/tmp/done.log".to_string()),
        ..Default::default()
    });
    record.complete(0);
    store.save(&record).unwrap();

    let result = query_status(Some(&store), "issue-105-json-executed", Some("json"));
    assert!(result.success);
    let output = result.output.unwrap();
    let parsed: serde_json::Value = serde_json::from_str(&output).unwrap();

    assert_eq!(parsed["status"], "executed");
    assert!(
        parsed.get("currentTime").is_none() || parsed["currentTime"].is_null(),
        "currentTime must not be present on completed records, got: {}",
        output
    );
}

#[test]
fn test_query_status_links_notation_includes_current_time_for_executing() {
    let (_temp_dir, store) = create_test_store();

    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 100".to_string(),
        uuid: Some("issue-105-links-executing".to_string()),
        pid: Some(99999),
        status: Some(ExecutionStatus::Executing),
        log_path: Some("/tmp/executing.log".to_string()),
        ..Default::default()
    });
    store.save(&record).unwrap();

    let result = query_status(Some(&store), "issue-105-links-executing", None);
    assert!(result.success);
    let output = result.output.unwrap();

    assert!(output.contains("status executing"));
    // currentTime should appear as an indented property with an ISO-like timestamp value
    let re = regex::Regex::new(r"\n  currentTime .*\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}").unwrap();
    assert!(
        re.is_match(&output),
        "currentTime missing or unrecognized in links-notation output: {}",
        output
    );
}

#[test]
fn test_query_status_text_includes_current_time_for_executing() {
    let (_temp_dir, store) = create_test_store();

    let record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "sleep 100".to_string(),
        uuid: Some("issue-105-text-executing".to_string()),
        pid: Some(99999),
        status: Some(ExecutionStatus::Executing),
        log_path: Some("/tmp/executing.log".to_string()),
        ..Default::default()
    });
    store.save(&record).unwrap();

    let result = query_status(Some(&store), "issue-105-text-executing", Some("text"));
    assert!(result.success);
    let output = result.output.unwrap();

    assert!(output.contains("Status:"));
    assert!(output.contains("executing"));
    assert!(output.contains("Current Time:"));
    // Current Time should appear right after Start Time
    let start_idx = output.find("Start Time:").expect("Start Time line");
    let current_idx = output.find("Current Time:").expect("Current Time line");
    assert!(current_idx > start_idx);
}

#[test]
fn test_query_status_text_omits_current_time_for_executed() {
    let (_temp_dir, store) = create_test_store();

    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "echo done".to_string(),
        uuid: Some("issue-105-text-executed".to_string()),
        pid: Some(11111),
        log_path: Some("/tmp/done.log".to_string()),
        ..Default::default()
    });
    record.complete(0);
    store.save(&record).unwrap();

    let result = query_status(Some(&store), "issue-105-text-executed", Some("text"));
    assert!(result.success);
    let output = result.output.unwrap();

    assert!(!output.contains("Current Time:"));
}

File: js/test/session-name-status.test.js
/**
 * Tests for --status lookup by session name and detached status enrichment
 * Issue #101: --session name not usable with --status, and --detached reports immediate completion
 */

const { describe, it, expect, beforeEach, afterEach } = require('bun:test');
const { spawnSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

const {
  ExecutionStore,
  ExecutionRecord,
  ExecutionStatus,
} = require('../src/lib/execution-store');
const {
  queryStatus,
  isDetachedSessionAlive,
  enrichDetachedStatus,
  attachCurrentTime,
} = require('../src/lib/status-formatter');

// Use temp directory for tests
const TEST_APP_FOLDER = path.join(
  os.tmpdir(),
  `session-name-status-test-${Date.now()}`
);

// Path to CLI
const CLI_PATH = path.join(__dirname, '../src/bin/cli.js');

// Helper to clean up test directory
function cleanupTestDir() {
  if (fs.existsSync(TEST_APP_FOLDER)) {
    fs.rmSync(TEST_APP_FOLDER, { recursive: true, force: true });
  }
}

// Helper to run CLI command
function runCli(args, env = {}) {
  const result = spawnSync('bun', [CLI_PATH, ...args], {
    encoding: 'utf8',
    env: {
      ...process.env,
      START_APP_FOLDER: TEST_APP_FOLDER,
      ...env,
    },
    timeout: 10000,
  });
  return {
    stdout: result.stdout || '',
    stderr: result.stderr || '',
    exitCode: result.status,
  };
}

describe('Issue #101: --status with session name lookup', () => {
  let store;

  beforeEach(() => {
    cleanupTestDir();
    store = new ExecutionStore({
      appFolder: TEST_APP_FOLDER,
      useLinks: false,
    });
  });

  afterEach(() => {
    cleanupTestDir();
  });

  describe('ExecutionStore.get() session name lookup', () => {
    it('should find record by UUID', () => {
      const record = new ExecutionRecord({
        command: 'echo hello',
        pid: 12345,
        logPath: '/tmp/test.log',
        options: { sessionName: 'my-custom-session' },
      });
      record.complete(0);
      store.save(record);

      const found = store.get(record.uuid);
      expect(found).not.toBeNull();
      expect(found.uuid).toBe(record.uuid);
    });

    it('should find record by session name', () => {
      const record = new ExecutionRecord({
        command: 'sleep 60',
        pid: 12345,
        logPath: '/tmp/test.log',
        options: {
          sessionName: 'my-custom-session',
          isolated: 'screen',
          isolationMode: 'detached',
        },
      });
      store.save(record);

      const found = store.get('my-custom-session');
      expect(found).not.toBeNull();
      expect(found.uuid).toBe(record.uuid);
      expect(found.options.sessionName).toBe('my-custom-session');
    });

    it('should prefer UUID match over session name', () => {
      // Create two records: one whose session name matches the UUID of another
      const record1 = new ExecutionRecord({
        command: 'echo first',
        pid: 111,
        logPath: '/tmp/first.log',
        options: { sessionName: 'some-session' },
      });
      record1.complete(0);
      store.save(record1);

      const record2 = new ExecutionRecord({
        command: 'echo second',
        pid: 222,
        logPath: '/tmp/second.log',
        options: { sessionName: record1.uuid }, // session name matches record1's UUID
      });
      store.save(record2);

      // Looking up by record1's UUID should return record1, not record2
      const found = store.get(record1.uuid);
      expect(found.command).toBe('echo first');
    });

    it('should return null for non-existent session name', () => {
      const found = store.get('nonexistent-session');
      expect(found).toBeNull();
    });

    it('should return null for record without session name', () => {
      const record = new ExecutionRecord({
        command: 'echo hello',
        pid: 12345,
        logPath: '/tmp/test.log',
      });
      store.save(record);

      const found = store.get('some-session-name');
      expect(found).toBeNull();
    });
  });

  describe('queryStatus() with session name', () => {
    it('should query status by session name', () => {
      const record = new ExecutionRecord({
        command: 'sleep 60',
        pid: 12345,
        logPath: '/tmp/test.log',
        options: {
          sessionName: 'my-test-session',
          isolated: 'screen',
          isolationMode: 'attached',
        },
      });
      record.complete(0);
      store.save(record);

      const result = queryStatus(store, 'my-test-session', 'json');
      expect(result.success).toBe(true);

      const parsed = JSON.parse(result.output);
      expect(parsed.uuid).toBe(record.uuid);
      expect(parsed.command).toBe('sleep 60');
    });

    it('should show error for non-existent session name', () => {
      const result = queryStatus(store, 'nonexistent-session', 'json');
      expect(result.success).toBe(false);
      expect(result.error).toContain('No execution found');
      expect(result.error).toContain('nonexistent-session');
    });
  });

  describe('CLI --status with session name', () => {
    it('should query by session name via CLI', () => {
      const record = new ExecutionRecord({
        command: 'echo hello world',
        pid: 12345,
        logPath: '/tmp/test.log',
        options: {
          sessionName: 'cli-test-session',
          isolated: 'screen',
          isolationMode: 'attached',
        },
      });
      record.complete(0);
      store.save(record);

      const result = runCli([
        '--status',
        'cli-test-session',
        '--output-format',
        'json',
      ]);

      expect(result.exitCode).toBe(0);
      const parsed = JSON.parse(result.stdout);
      expect(parsed.uuid).toBe(record.uuid);
      expect(parsed.command).toBe('echo hello world');
      expect(parsed.status).toBe('executed');
    });

    it('should show error for non-existent session name via CLI', () => {
      const result = runCli(['--status', 'nonexistent-session']);
      expect(result.exitCode).toBe(1);
      expect(result.stderr).toContain('No execution found');
    });
  });
});

describe('Issue #101: Detached status enrichment', () => {
  describe('isDetachedSessionAlive()', () => {
    it('should return null for non-detached records', () => {
      const record = new ExecutionRecord({
        command: 'echo hello',
        options: {
          sessionName: 'test',
          isolated: 'screen',
          isolationMode: 'attached',
        },
      });
      expect(isDetachedSessionAlive(record)).toBeNull();
    });

    it('should return null for records without session name', () => {
      const record = new ExecutionRecord({
        command: 'echo hello',
        options: { isolationMode: 'detached' },
      });
      expect(isDetachedSessionAlive(record)).toBeNull();
    });

    it('should return false for non-existent screen session', () => {
      const record = new ExecutionRecord({
        command: 'sleep 60',
        options: {
          sessionName: 'nonexistent-screen-session-test-101',
          isolated: 'screen',
          isolationMode: 'detached',
        },
      });
      const alive = isDetachedSessionAlive(record);
      // May be false or null depending on whether screen is installed
      if (alive !== null) {
        expect(alive).toBe(false);
      }
    });

    it('should return false for non-existent docker container', () => {
      const record = new ExecutionRecord({
        command: 'sleep 60',
        options: {
          sessionName: 'nonexistent-docker-container-test-101',
          isolated: 'docker',
          isolationMode: 'detached',
        },
      });
      const alive = isDetachedSessionAlive(record);
      // May be false or null depending on whether docker is installed
      if (alive !== null) {
        expect(alive).toBe(false);
      }
    });
  });

  describe('enrichDetachedStatus()', () => {
    it('should not modify non-detached records', () => {
      const record = new ExecutionRecord({
        command: 'echo hello',
        options: {
          sessionName: 'test',
          isolated: 'screen',
          isolationMode: 'attached',
        },
      });
      record.complete(0);

      const enriched = enrichDetachedStatus(record);
      expect(enriched.status).toBe('executed');
      expect(enriched.exitCode).toBe(0);
    });

    it('should mark non-running detached session as executed', () => {
      const record = new ExecutionRecord({
        command: 'sleep 60',
        options: {
          sessionName: 'nonexistent-session-enrich-test-101',
          isolated: 'screen',
          isolationMode: 'detached',
        },
      });
      // Record says executing, but session doesn't exist

      const enriched = enrichDetachedStatus(record);
      // If screen is available, should mark as executed
      // If screen is not available, should return unchanged
      if (enriched.status === 'executed') {
        expect(enriched.exitCode).toBe(-1); // Unknown exit code
        expect(enriched.endTime).not.toBeNull();
      }
    });
  });
});

describe('Issue #105: attachCurrentTime for executing status', () => {
  it('should add currentTime to serialization when status is executing', () => {
    const record = new ExecutionRecord({
      command: 'sleep 60',
      pid: 12345,
      logPath: '/tmp/test.log',
    });

    const before = Date.now();
    const wrapped = attachCurrentTime(record);
    const obj = wrapped.toObject();
    const after = Date.now();

    expect(obj.currentTime).toBeDefined();
    const currentTimeMs = new Date(obj.currentTime).getTime();
    expect(Number.isNaN(currentTimeMs)).toBe(false);
    expect(currentTimeMs).toBeGreaterThanOrEqual(before - 1);
    expect(currentTimeMs).toBeLessThanOrEqual(after + 1);
  });

  it('should not add currentTime when status is executed', () => {
    const record = new ExecutionRecord({
      command: 'echo hello',
      pid: 12345,
      logPath: '/tmp/test.log',
    });
    record.complete(0);

    const wrapped = attachCurrentTime(record);
    // attachCurrentTime should return the original record unchanged
    expect(wrapped).toBe(record);
    const obj = wrapped.toObject();
    expect(obj.currentTime).toBeUndefined();
  });

  it('should not mutate the original record', () => {
    const record = new ExecutionRecord({
      command: 'sleep 60',
      pid: 12345,
      logPath: '/tmp/test.log',
    });

    const wrapped = attachCurrentTime(record);
    expect(wrapped).not.toBe(record);
    // The original record's toObject output should not include currentTime
    const originalObj = record.toObject();
    expect(originalObj.currentTime).toBeUndefined();
  });

  it('should place currentTime right after startTime in serialization order', () => {
    const record = new ExecutionRecord({
      command: 'sleep 60',
      pid: 12345,
      logPath: '/tmp/test.log',
    });

    const wrapped = attachCurrentTime(record);
    const keys = Object.keys(wrapped.toObject());
    const startIndex = keys.indexOf('startTime');
    expect(startIndex).toBeGreaterThanOrEqual(0);
    expect(keys[startIndex + 1]).toBe('currentTime');
  });

  it('should handle null record gracefully', () => {
    expect(attachCurrentTime(null)).toBeNull();
  });
});

console.log('=== Session Name Status Tests (Issue #101) ===');
